### ISSUE TITLE
cannot specialize to this type

### ISSUE THREAD

**Issue by dopefishh** [clean-cloogle/common-problems#50]
**Exact error message**:

> cannot specialize to this type

**A minimal example**:

The generics system can only derive generic functions for first-order types.
If a derivation for a higher order type is tried it throws this error.

```clean
module test
import StdGeneric

generic g a :: a
g{|OBJECT|} g = OBJECT g
g{|CONS|} g = CONS g

derive g T, S, Int

//kind *->*
:: S a = S a

//kind (*->*)->*
:: T a = T (a Int)

Start :: T S
Start = g{|*|}
```

**Comment by camilstaps** [comment]
The error here is that the instance for Int cannot be derived, because this is a basic type. There is no problem with the other types.

**Comment by dopefishh** [comment]
Hmm. There I was thinking you cannot derive for higher order types. I wonder what they mean by the statement on page 8 then: https://repository.ubn.ru.nl/bitstream/handle/2066/103802/103802.pdf?sequence=1

**Comment by dopefishh** [comment]
It still is an undocumented error, that's the upside...

**Comment by camilstaps** [comment]
Presumably it signals that there cannot be editors for higher-order types, i.e., there is no editor for `Maybe`, although there is an editor for `Maybe t` where `t :: *`. It is unclear what such an editor would do anyway since in Clean all values are of first order types.

**Comment by dopefishh** [comment]
My T is, according te the afp slides
According to the interwebs Maybe is a higher order type.
I can't find conclusive literature on it and Maybe not.
Higher-ranked, Higher-order, Higher-kinded 
According to https://stackoverflow.com/questions/38003467/disambiguation-of-higher-kinded-type-vs-higher-order-type order/kind are the same.

**Comment by camilstaps** [comment]
Resolved in #55.
